repr,expr,sql: remove ScalarType::Unknown

benesch opened the pull request:
@justinj this fixes #3150 and as a side effect cleans up a long-standing bit of tech debt.

@sploiselle I think we chatted about this a bit last week. Don't worry: I promise to wait until after your function generalization PR! But I think it has the potential to clean up some of the weirdness around `param_types`, since it exposes a function (`plan_coercible_expr`) that plans an expression without mutating `param_types`. And now that I look at it, with this PR and your PR I think literal coercion (#481) will wind up being like <10 lines of code.

/cc @frankmcsherry I think you'll be delighted about the effects on the expr/dataflow layer.

----

ScalarType::Unknown is a Postgresism that has not served us well. It
is best understood as the type of NULL in the following SQL statement

    SELECT NULL

although in recent versions Postgres will assign that NULL a type of
`text` before planning is complete. That is in fact the key insight: the
"unknown" type represents a transient state of the planner. Either the
unknown type will become known by the time planning is complete, or it
will not, and planning the query will fail.

We can isolate the dataflow execution layer from the unknown type by
limiting its existence to the few moments in the planner where it is
needed. To that end, this patch removes the ScalarType::Unknown variant,
and introduces a CoercibleScalarExpr to take its place.

The idea is that the planner has special handling for expressions which
could possibly coerce to another type (i.e., a string literal, list
literal, NULL literal, or query parameter). The expressions will be
coerced at the moment their true types are required; for example calling
a function will coerce any unknown parameters, as will reaching the root
of the expression tree.

All code outside of the `sql` crate gets vastly simpler. The SQL planner
gets simpler in many places too, and a bit different in others--not more
complicated, necessarily, just different.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/3160)
<!-- Reviewable:end -->

benesch commented on src/sql/expr.rs:
```diff
@@ -145,6 +145,46 @@ pub enum ScalarExpr {
+    Parameter(usize),
+    LiteralNull,
+    LiteralList(Vec<CoercibleScalarExpr>),
+}
```
Yeah exactly! I didn't include it here to keep clippy happy, but basically you just introduce a `CoercibleScalarExpr::LiteralString(String)`, unconditionally plan string literals as such, and the rest falls out pretty naturally. Like I think you just implement `plan_coerce` for `CoercibleScalarExpr::LiteralString` by calling `plan_cast_internal()` with the desired type.

benesch replied:
This is done now!

sploiselle commented on src/sql/src/func.rs:
```diff
@@ -587,52 +584,49 @@ impl<'a> ArgImplementationMatcher<'a> {
+            }
+        }
+
+        if ParamType::JsonbAny == *typ {
```
This could be a `match typ` now that we're no longer comparig `arg.is_string_literal()`

benesch replied:
Sure thing, done.

sploiselle commented on src/sql/src/func.rs:
```diff
@@ -254,22 +252,30 @@ impl<'a> ArgImplementationMatcher<'a> {
-            // Best match
-            m.best_match(args, &raw_arg_types)?
-        };
+        let f = m.best_match(&exprs)?;
```
without an attempt at an exact match, `best_match` seems misnamed

benesch replied:
Shoot, I forgot I removed that. I guess it's not strictly necessary given that everything passed without it, but I added it back so that we could follow Postgres's algorithm exactly.

sploiselle commented on src/sql/src/query.rs:
```diff
@@ -1799,9 +1846,9 @@ fn plan_aggregate(ecx: &ExprContext, sql_func: &Function) -> Result<AggregateExp
-            // can't ever escape the surrounding reduce.
-            ScalarExpr::literal_null(ScalarType::Unknown),
+            // The scalar type of the null doesn't matter because this
+            // expression can't ever escape the surrounding reduce.
```
nit: s/can't ever escape/can't escape/ ...or something grammatically equivalent

benesch replied:
Done, but I"m not sure I see what was wrong with the original?

sploiselle commented on src/sql/src/query.rs:
```diff
@@ -3076,7 +3109,6 @@ pub fn plan_cast_implicit<'a>(
     let from_scalar_type = ecx.column_type(&expr).scalar_type;
 
     match (&from_scalar_type, &to_scalar_type) {
-        (String, String) | (Unknown, String) => {}
```
You'll need to re-add this `(String, String)` line for the time being; I have a refactor for the implicit casts in mind but will wait for this to land.

benesch replied:
Oh crap. Did not mean to delete the `(String, String)` bit.

benesch commented on src/sql/expr.rs:
```diff
@@ -145,6 +145,46 @@ pub enum ScalarExpr {
     Select(Box<RelationExpr>),
 }
 
+pub enum CoercibleScalarExpr {
```
Done!

benesch commented:
Thanks for dealing with the sloppiness of the revision you saw, @sploiselle! I've cleaned things in func.rs up a good bit, and the set of changes is smaller. Nothing fundamental has shifted, though; just a few more comments and a cleaner separation of responsibility between `coerce_arg_to_type` and `plan_coerce`. I think probably those two functions want to be merged somehow, but I don't think that can happen until we've moved everything over to func.rs.
